src/Std/Sat/AIG/CNF.lean

```diff
@@ -0,0 +1,718 @@
+
+theorem Cache.IsExtensionBy_trans_left (cache1 : Cache aig cnf1) (cache2 : Cache aig cnf2)
+    (cache3 : Cache aig cnf3) (h12 : IsExtensionBy cache1 cache2 new1 hnew1)
+    (h23 : IsExtensionBy cache2 cache3 new2 hnew2) : IsExtensionBy cache1 cache3 new1 hnew1 := by
+  apply IsExtensionBy.mk
+  . intro idx hidx hmarked
+    apply h23.extension
+    . apply h12.extension
+      . exact hmarked
+      . omega
+    . omega
+  . apply h23.extension
+    . exact h12.trueAt
+    . omega
+
+theorem Cache.IsExtensionBy_trans_right (cache1 : Cache aig cnf1) (cache2 : Cache aig cnf2)
+    (cache3 : Cache aig cnf3) (h12 : IsExtensionBy cache1 cache2 new1 hnew1)
+    (h23 : IsExtensionBy cache2 cache3 new2 hnew2) : IsExtensionBy cache1 cache3 new2 hnew2 := by
+  apply IsExtensionBy.mk
+  . intro idx hidx hmarked
+    apply h23.extension
+    . apply h12.extension
+      . exact hmarked
+      . omega
+    . omega
+  . exact h23.trueAt
+
+/--
+Cache extension is a reflexive relation.
+-/
+theorem Cache.IsExtensionBy_rfl (cache : Cache aig cnf) {h} (hmarked : cache.marks[idx]'h = true) :
+    Cache.IsExtensionBy cache cache idx (have := cache.hmarks; omega) := by
+  apply IsExtensionBy.mk
+  . intros
+    assumption
+  . exact hmarked
+
+theorem Cache.IsExtensionBy_set (cache1 : Cache aig cnf1) (cache2 : Cache aig cnf2) (idx : Nat)
+    (hbound : idx < cache1.marks.size) (h : cache2.marks = cache1.marks.set ⟨idx, hbound⟩ true) :
+    IsExtensionBy cache1 cache2 idx (by have := cache1.hmarks; omega) := by
+  apply IsExtensionBy.mk
+  . intro idx hidx hmark
+    simp [Array.getElem_set, hmark, h]
+  . simp[h]
+
+/--
+A cache with no entries is valid for an empty CNF.
+-/
+def Cache.init (aig : AIG Nat) : Cache aig [] where
+  marks := mkArray aig.decls.size false
+  hmarks := by simp
+  inv := Inv_init
+
+/--
+Add a `Decl.const` to a `Cache`.
+-/
+def Cache.addConst (cache : Cache aig cnf) (idx : Nat) (h : idx < aig.decls.size)
+    (htip : aig.decls[idx]'h = .const b) :
+    {
+      out : Cache aig (Decl.constToCNF (.inr ⟨idx, h⟩) b ++ cnf)
+        //
+      Cache.IsExtensionBy cache out idx h
+    } :=
+  have hmarkbound : idx < cache.marks.size := by have := cache.hmarks; omega
+  let out :=
+    { cache with
+      marks := cache.marks.set ⟨idx, hmarkbound⟩ true
+      hmarks := by simp [cache.hmarks]
+      inv := by
+        constructor
+        . intro lhs rhs linv rinv idx hbound hmarked heq
+          rw [Array.getElem_set] at hmarked
+          split at hmarked
+          . simp_all
+          . have := cache.inv.hmark lhs rhs linv rinv idx hbound hmarked heq
+            simp [Array.getElem_set, this]
+        . intro assign heval idx hbound hmarked
+          rw [Array.getElem_set] at hmarked
+          split at hmarked
+          . next heq =>
+            simp at heq
+            simp [heq] at htip heval
+            simp [heval, denote_idx_const htip]
+          . next heq =>
+            simp at heval
```

Comment:
non-terminal simp?

Reply:
I think the pattern of 

```
simp at heq
simp [heq] at htip heval
```

occurs in every proof that shows that `inv` is preserved. and they all appear to be non-terminal to me.